bpython/curtsiesfrontend/repl.py

```diff
@@ -1076,12 +1167,18 @@ def add_normal_character(self, char):
                 reset_rl_history=False,
                 clear_special_mode=False,
             )
-            self.cursor_offset += 1
+            if add_to_search:
+                self.cursor_offset += 1
+            else:
+                self._cursor_offset += 1
```

Comment:
This doesn't work for me, and I think it could be fixed or just removed instead of fixed because of how complicated it is. I was hoping that typing

```
>>> def foo(
```
would still autocomplete a previous
```
>>> def foo(a, b, c):
```
but that doesn't work for me with this code.

Instead, I think it's more important to fix the explicit reverse history search mode. Try explicit reverse history search, by default it's option-r.
![image](https://user-images.githubusercontent.com/458879/140094663-f7febf3f-2678-4602-9092-28e176d255c3.png)

Reply:
so should i remove this and focus on the -r option?

Reply:
btw, when you type `def foo(` , you see the hint right? i got it there but couldnt make it to complete the line properly

Reply:
oh huh, no I couldn't get it to show up. Maybe we have some other different settings? or maybe it's not clear what I was thinking, here's a video of it "not working" for me (it's a pretty minor thing)

https://user-images.githubusercontent.com/458879/140108798-165abd56-d0e4-41a0-a75d-58903ee0414b.mov

Reply:
yes. this is behavior on my side as well. i though the history did not even show for you in the box. before i added the conditions(in code quoted in this comment), it did not even show up. i can try to make it work. what is the requested behavior in the reverse search? i havent used it that much.

Reply:
in terms of incremental search, i wanted to make it so the hint will show up in the box (as it does now), and upon pressing TAB it will complete. I think the problem is inside the on_tab, which makes the "move cursor behind closing paren" trigger before completion has a change. i will look into it

Reply:
Ah got it, that makes sense — the hint showing up is way more important, good fix. I don't think the history thing (not being shown in grey due to the closing paren) I pointed out is worth fixing, and if it is, we can fix by changing the rl_history search algorithm.

The explicit (invoked by option-r) incremental search issue I was worried about is the extra close paren here:

https://user-images.githubusercontent.com/458879/140129899-52933dc1-3153-43fe-91fe-47ee984f89eb.mov

Reply:
Yeah that sounds good to me, I was confused about what the fix you had did when I suggested removing it.